src/main/java/za/co/neslotech/spring/trainspringbootinsight/controller/AGenericController.java

```diff
@@ -0,0 +1,69 @@
+package za.co.neslotech.spring.trainspringbootinsight.controller;
+
+import jakarta.persistence.EntityNotFoundException;
+import org.springframework.beans.BeanUtils;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+import java.util.List;
+
+/**
+ * A generic REST API Controller with CRUD operations.
+ * The child class should use the following annotations
+ *      1. @RestController
+ * @param <T> The entity class.
+ * @param <I> The primary key field type.
+ */
+public abstract class AGenericController <T, I> {
+
+    protected abstract JpaRepository<T, I> getRepository();
+    protected abstract String getEntityNotFoundDescription(final I id);
+    protected abstract String[] getUpdateColumnExclusions();
+
+    @GetMapping
+    public List<T> list(){
+        return getRepository().findAll();
```

Comment:
The good practice pattern we follow in RESTful API's is to have a controller -> service -> repository flow.
This allows us to move any logic specific to the entity in question into a service bean that can be injected into other services without having to rewrite the logic

Reply:
Hi Jordan, I am not sure I understand this part. Should there be a service between the repository and the controller?